### DISCUSSION TITLE
Cross-compiling and <dlfcn.h>

### DISCUSSION THREAD

**Discussion by Bossach** [pocl/pocl#1417] [General]
I had some problems when cross-compiling PoCL (for another non-x86 linux system). There was some "implicit declaration" of "`dlopen()`, `dlclose()` .." and so on.

It is caused by not including `<dlfcn.h>` hided under `#ifdef`

For example:
```
/lib/CL/devices/devices.c:115
...
#ifdef HAVE_DLFCN_H
#if defined(__APPLE__)
#define _DARWIN_C_SOURCE
#endif
#include <dlfcn.h>
#endif
...
```

After some digging i found `HAVE_DLFCN_H` sets here under some conditions:
```
/CMakeLists.txt:964
...
if (CMAKE_CROSSCOMPILING AND (NOT ENABLE_HOST_CPU_DEVICES) AND (NOT ENABLE_HSA))
    message(STATUS "Cross-compiling without CPU/HSA devices -> skipping LIBDL search")
  else()
    ...
       set(HAVE_DLFCN_H ON CACHE BOOL "dlfcn.h" FORCE)
    ...
...
```

So temporary fix is build with `-DENABLE_HOST_CPU_DEVICES=ON` and it compiles and all works just fine.
But i don't need HOST_CPU_DEVICE neither HSA, just wanna build with my device driver only.

So the questions:

1) What the reason to hide finding and defining `dlfcn` under such condition?
(`if (CMAKE_CROSSCOMPILING AND (NOT ENABLE_HOST_CPU_DEVICES) AND (NOT ENABLE_HSA))`)
(I assume that may works just for me, but in other cases this condition may have sense)

2) How it intended to compile in case of not defined `HAVE_DLFCN_H` without including `<dlfcn.h>` despite that, for example, `devices.c` contains related function calls?  (`dlopen()`, `dlclose()` and so on)

Thanks!

**Comment by franz** [comment]
Hello,

> What the reason to hide finding and defining dlfcn under such condition?

The reason was to skip searching for `dlfcn.h` when it's not needed and might not work (e.g. cross-compiling for Android). Originally the only code using `dlopen` was in the CPU drivers and the HSA driver.

> despite that, for example, devices.c contains related function calls

The code in `devices.c` that uses `dlopen` is ifdef'ed on `ENABLE_LOADABLE_DRIVERS`, the problem is that the CMake if command was not update to reflect this. If you compile without loadable drivers and without CPU/HSA then dlfcn should not be required at all.

**Reply by Bossach** [reply]
Thank you for your answer!

Indeed, i successfuly compiled it with just my device using `-DENABLE_LOADABLE_DRIVERS=OFF`, thanks for the tip.
_(despite now something brokes at runtime (some xxxShedulerxxx symbol missing ), maybe it related to my custom device driver, will dig for it later)_

So, as i understand, the updated "CMake if command" like `if (CMAKE_CROSSCOMPILING AND (NOT ENABLE_HOST_CPU_DEVICES) AND (NOT ENABLE_HSA) AND (NOT ENABLE_LOADABLE_DRIVERS))` will be OK?
_(and/or theoretically `if( ... AND (NOT ENABLE_SOME_DL_USING_DRIVER))`)_
Did i understand correctly? (Or it's not OK because will break something else)

Thanks!

**Reply by franz** [reply]
Hi, sorry for late reply. Yeah updating the CMake line like that should be OK, probably ENABLE_PROXY_DEVICE also needs to be added into the if condition (it seems proxy driver includes `dlfcn.h`, but i haven't examined the code in detail).
